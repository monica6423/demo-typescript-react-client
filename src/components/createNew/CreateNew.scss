@import "../common.scss";

.flex {
  display: flex;
}
.search-box{
  width: 300px !important;
  max-width: 900px !important;
}
td.search-box {
  position: relative;

  input {
    height: 42px;
    padding: 12px 20px;
    box-sizing: border-box;
    background-color: #fafafa;
    font-size: 13px;
    border: 1px solid #eeeeee;
    border-radius: 5px;
    font-weight: 500;
    margin-left: 16px;
    width: 285px;
  }
}

.button-cell {
  height: 100%;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  width: fit-content;
  .button {
    height: 40px;
    background-color: #ededed;
    border: none;
    color: $secondary;
    font-weight: 500;
    padding: 10px 16px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 16px;
    margin: 4px 2px;
    cursor: pointer;
    margin-right: 16px;
    &.cancel-button {
      color: $button;
    }
    &.save-button {
      background-color: $button;
      color: white;
    }
  }
}

.main {
  background-color:#acb5bd !important;
  color: antiquewhite !important;
}

.error-message {
  color: #cc0033;
  display: inline-block;
  font-size: 12px;
  line-height: 15px;
  margin: 5px 0 0;
  position: absolute;
  bottom: 0;
  right: 10px;
}
