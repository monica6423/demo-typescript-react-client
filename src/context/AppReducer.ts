import { Restaurant, Company, RestaurantStatus } from "../interfaces/";

export type Action =
  | {
      type: "FETCH_RESTAURANTS";
      payload: Restaurant[];
    }
  | {
      type: "FETCH_COMPANIES";
      payload: Company[];
    }
  | {
      type: "FETCH_PARENT_COMPANIES";
      payload: Company[];
    }
  | {
      type: "FETCH_STATION_BY_ID";
      payload: Restaurant;
    }
  | {
      type: "FETCH_RESTAURANTS_BY_COMPANY_ID";
      payload: Restaurant[];
    }
  | {
      type: "FETCH_STATION_TYPE_BY_ID";
      payload: Restaurant;
    }
  | {
      type: "FETCH_STATION_TYPES";
      payload: RestaurantStatus[];
    }
  | {
      type: "FETCH_COMPANY_BY_ID";
      payload: Company;
    }
  | {
      type: "ADD_STATION";
      payload: Restaurant;
    }
  | {
      type: "ADD_COMPANY";
      payload: Company;
    };

const appReducer = (state: any, action: any) => {
  switch (action.type) {
    case "FETCH_RESTAURANTS":
      return {
        ...state,
        restaurants: [...action.payload],
      };
    case "FETCH_COMPANIES":
      return {
        ...state,
        companies: [...action.payload],
      };
    case "FETCH_PARENT_COMPANIES":
      return {
        ...state,
        parentCompanies: [...action.payload],
      };
    case "FETCH_STATION_BY_ID":
      return {
        ...state,
        restaurant: action.payload,
      };
    case "FETCH_RESTAURANTS_BY_COMPANY_ID":
      return {
        ...state,
        restaurantById: action.payload,
      };
    case "FETCH_STATION_TYPE_BY_ID":
      return {
        ...state,
        restaurantType: action.payload,
      };
    case "FETCH_STATION_TYPES":
      return {
        ...state,
        restaurantTypes: action.payload,
      };
    case "FETCH_COMPANY_BY_ID":
      return {
        ...state,
        companyById: action.payload,
      };
    case "ADD_STATION":
      return {
        ...state,
        restaurants: [action.payload, ...state.restaurants],
      };
    case "ADD_COMPANY":
      return {
        ...state,
        companies: [action.payload, ...state.companies],
      };
    default:
      return state;
  }
};

export default appReducer;
