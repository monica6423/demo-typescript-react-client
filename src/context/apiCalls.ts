import { Restaurant, Company, RestaurantType } from "../interfaces/";
const apiHost = process.env.REACT_APP_API_HOST;

export const createCompany = async (data: Company) => {
  try {
    const res = await fetch(`${apiHost}create-company`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
    return res.json();
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const createRestaurant = async (data: Restaurant) => {
  try {
    const res = await fetch(`${apiHost}create-restaurant`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
    return res.json();
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const createRestaurantType = async (data: RestaurantType) => {
  try {
    const res = await fetch(`${apiHost}create-restaurant-type`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
    return res.json();
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const fetchRestaurant = async () => {
  try {
    const res = await fetch(`${apiHost}get-restaurants`);
    return await res.json();
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const fetchRestaurantById = async (id: string) => {
  try {
    const res = await fetch(`${apiHost}get-restaurant?id=${id}`);
    return await res.json();
  } catch (error) {
    console.error(error);
    throw error;
  }
};
