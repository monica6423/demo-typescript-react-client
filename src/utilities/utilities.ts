export const debounce = (fn: any, delay = 300) => {
  let timer: any;
  return (...args: string[]) => {
    if (timer) {
      clearTimeout(timer);
    }
    timer = setTimeout(() => {
      timer = null;
      fn(...args);
    }, delay);
  };
};
