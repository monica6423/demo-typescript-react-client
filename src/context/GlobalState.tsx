import React, {
  createContext,
  useReducer,
  ReactElement,
  useEffect,
  useCallback,
} from "react";
import AppReducer, { Action } from "./AppReducer";
import {
  Restaurant,
  Company,
  RestaurantByCompany,
  RestaurantType,
} from "../interfaces/";
import {
  createCompany,
  createRestaurant,
  createRestaurantType,
  fetchRestaurant,
  fetchRestaurantById,
} from "./apiCalls";

const apiHost = process.env.REACT_APP_API_HOST;

interface InitialState {
  restaurants: Restaurant[] | [];
  restaurant: Restaurant | null;
  createData: (
    type: string,
    data: Company | Restaurant | RestaurantType
  ) => void;
  getRestaurant: () => void;
  getRestaurantById: (id: string) => void;
  editRestaurant: (restaurant: Restaurant) => void;
  getRestaurantTypeById: (id: string) => void;
  restaurantType: string | null;
  editRestaurantType: (data: RestaurantType) => void;
  companies: Company[];
  parentCompanies: Company[];
  getRestaurantsByCompanyId: (id: string) => void;
  restaurantById: RestaurantByCompany[];
  restaurantTypes: RestaurantType[] | [];
  getCompanies: (value: boolean) => Promise<void>;
  dispatch: React.Dispatch<Action>;
  getCompanyById: (id: string) => void;
  companyById: Company[];
  editCompany: (data: Company) => void;
}
// Initial state
const initialState: InitialState = {
  restaurants: [],
  restaurant: null,
  createData: () => {},
  getRestaurant: () => {},
  getRestaurantById: () => {},
  editRestaurant: () => {},
  getRestaurantTypeById: () => {},
  restaurantType: null,
  editRestaurantType: () => {},
  companies: [],
  parentCompanies: [],
  getRestaurantsByCompanyId: () => {},
  restaurantById: [],
  restaurantTypes: [],
  getCompanies: () => {
    return Promise.resolve();
  },
  dispatch: () => {},
  getCompanyById: () => {},
  companyById: [],
  editCompany: () => {},
};

// Create context
export const GlobalContext = createContext(initialState);

// Provider component
interface GlobalProviderProps {
  children?: ReactElement[] | ReactElement;
}
export const GlobalProvider = ({ children }: GlobalProviderProps) => {
  const [state, dispatch] = useReducer(AppReducer, initialState);

  useEffect(() => {
    getRestaurant();
    getCompanies(true);
    getCompanies();
    getRestaurantTypes();
  }, []);

  // Actions
  const createData = async (
    type: string,
    data: Company | Restaurant | RestaurantType
  ) => {
    switch (type) {
      case "company":
        const company = await createCompany(data as Company);
        dispatch({
          type: "ADD_COMPANY",
          payload: company,
        });
        break;
      case "restaurant":
        const restaurant = await createRestaurant(data as Restaurant);
        const newRestaurant = {
          restaurant: restaurant.name,
          company: restaurant.company.name,
          restaurantType: restaurant.restaurantType.name,
          restaurantTypeId: restaurant.restaurantType.id,
          id: restaurant.id,
          companyId: restaurant.company.id,
        };
        dispatch({
          type: "ADD_RESTAURANT",
          payload: newRestaurant,
        });
        break;
      case "restaurantType":
        const restaurantType = await createRestaurantType(
          data as RestaurantType
        );
        dispatch({
          type: "ADD_RESTAURANT_TYPES",
          payload: restaurantType,
        });
        break;
      default:
        break;
    }
  };

  const getRestaurant = async () => {
    const data = await fetchRestaurant();
    const restaurants = data.map((restaurant: any) => ({
      restaurant: restaurant.name,
      company: restaurant.company.name,
      restaurantType: restaurant.restaurantType.name,
      restaurantTypeId: restaurant.restaurantType.id,
      id: restaurant.id,
      companyId: restaurant.company.id,
    }));
    dispatch({
      type: "FETCH_RESTAURANTS",
      payload: restaurants,
    });
  };

  const getRestaurantById = useCallback(async (id: string) => {
    const data = await fetchRestaurantById(id);
    dispatch({
      type: "FETCH_STATION_BY_ID",
      payload: data,
    });
  }, []);

  const getRestaurantTypes = async () => {
    const res = await fetch(`${apiHost}get-restaurant-type`);
    const data = await res.json();
    dispatch({
      type: "FETCH_RESTAURANT_TYPES",
      payload: data,
    });
  };

  const getRestaurantTypeById = useCallback(async (id: string) => {
    const res = await fetch(`${apiHost}get-restaurant-type?id=${id}`);
    const data = await res.json();
    dispatch({
      type: "FETCH_RESTAURANT_TYPE_BY_ID",
      payload: data,
    });
  }, []);

  const editRestaurant = async (restaurant: Restaurant) => {
    await fetch(`${apiHost}create-restaurant`, {
      method: "post",
      body: JSON.stringify(restaurant),
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
    });
  };

  const editRestaurantType = async (data: any) => {
    await fetch(`${apiHost}create-restaurant-type`, {
      method: "post",
      body: JSON.stringify(data),
    });
  };

  const editCompany = async (data: any) => {
    await fetch(`${apiHost}create-company`, {
      method: "post",
      body: JSON.stringify(data),
    });
  };

  const getCompanies = async (value = false) => {
    const res = await fetch(`${apiHost}get-companies?parent=${value}`);
    const data = await res.json();
    if (value) {
      dispatch({
        type: "FETCH_PARENT_COMPANIES",
        payload: data,
      });
    }
    dispatch({
      type: "FETCH_COMPANIES",
      payload: data,
    });
  };

  const getRestaurantsByCompanyId = useCallback(async (id: string) => {
    const res = await fetch(`${apiHost}get-restaurants-by-companyId?id=${id}`);
    const data = await res.json();
    dispatch({
      type: "FETCH_RESTAURANTS_BY_COMPANY_ID",
      payload: data,
    });
  }, []);

  const getCompanyById = useCallback(async (id: string) => {
    const res = await fetch(`${apiHost}get-company-by-id?id=${id}`);
    const data = await res.json();
    dispatch({
      type: "FETCH_COMPANY_BY_ID",
      payload: data[0],
    });
  }, []);

  return (
    <GlobalContext.Provider
      value={{
        restaurants: state.restaurants,
        restaurant: state.restaurant,
        createData,
        getRestaurant,
        getRestaurantById,
        editRestaurant,
        getRestaurantTypeById,
        restaurantType: state.restaurantType,
        editRestaurantType,
        companies: state.companies,
        parentCompanies: state.parentCompanies,
        getRestaurantsByCompanyId,
        restaurantById: state.restaurantById,
        restaurantTypes: state.restaurantTypes,
        getCompanies,
        getCompanyById,
        companyById: state.companyById,
        editCompany,
        dispatch,
      }}
    >
      {children}
    </GlobalContext.Provider>
  );
};
