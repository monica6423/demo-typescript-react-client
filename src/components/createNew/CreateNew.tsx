import React, { useState, ChangeEvent, useContext, useMemo } from "react";
import { GlobalContext } from "../../context/GlobalState";
import "./CreateNew.scss";

interface setState {
  setFormType: React.Dispatch<React.SetStateAction<string | null>>;
}

const CreateNew = ({ setFormType }: setState) => {
  const { dispatch } = useContext(GlobalContext);
  const [open, setOpen] = useState<boolean>(false);
  const [searchTerm, setSearchTerm] = useState<string>("");
  const renderForm = (type: string) => {
    setFormType(type);
  };

  const debounce = (fn: any, delay = 300) => {
    let timer: any;
    return (...args: string[]) => {
      console.log(timer);
      console.log("fetchData2 inside return");

      if (timer) {
        console.log("clear");
        clearTimeout(timer);
      }
      console.log("timer", timer);
      timer = setTimeout(() => {
        console.log("settimeout");
        timer = null;
        fn(...args);
      }, delay);
    };
  };

  const fetchData = async (value: any) => {
    const res = await fetch(
      `http://localhost:3000/dev/api/get-restaurants?searchValue=${value}`
    );
    const data = await res.json();
    const restaurants = data.map((restaurant: any) => ({
      restaurant: restaurant.name,
      company: restaurant.company.name,
      restaurantType: restaurant.restaurantType.name,
      restaurantTypeId: restaurant.restaurantType.id,
      id: restaurant.id,
      companyId: restaurant.company.id,
    }));
    dispatch({
      type: "FETCH_RESTAURANTS",
      payload: restaurants,
    });
  };

  const debounceHandler = useMemo(() => debounce(fetchData, 1000), []);

  const onSearch = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(e.target.value);
    debounceHandler(e.target.value);
  };

  return (
    <div>
      <div className="button-cell ">
        <button className="button main" onClick={(e) => setOpen(!open)}>
          <b>Create new</b>
        </button>
        <td className="search-box">
          <input
            type="text"
            name={"inputField.key"}
            value={searchTerm}
            onChange={(e) => onSearch(e)}
            placeholder={"Search by restaurant, company & type"}
          ></input>
        </td>
      </div>
      {open && (
        <div className="flex button-cell">
          <button className="button " onClick={(e) => renderForm("restaurant")}>
            Restaurant
          </button>
          <button
            className="button "
            onClick={(e) => renderForm("restaurantType")}
          >
            Restaurant Type
          </button>
          <button className="button " onClick={(e) => renderForm("company")}>
            Company
          </button>
        </div>
      )}
    </div>
  );
};

export default CreateNew;
