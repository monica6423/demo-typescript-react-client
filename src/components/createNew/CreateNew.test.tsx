import React from "react";
import { render, fireEvent, cleanup, screen } from "@testing-library/react";
import CreateNew from "./CreateNew";
import { GlobalProvider } from "../../context/GlobalState";
import { debounce } from "../../utilities/utilities";

describe("CreateNew component", () => {
  afterEach(cleanup);

  it("should render the component", () => {
    const { asFragment } = render(
      <GlobalProvider>
        <CreateNew />
      </GlobalProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should open the create new modal when create new button is clicked", () => {
    const { getByText } = render(
      <GlobalProvider>
        <CreateNew />
      </GlobalProvider>
    );
    const createNewButton = getByText(/Create new/i);
    fireEvent.click(createNewButton);
    expect(createNewButton).toHaveTextContent("Create new");
    // You can add more assertions to check if the modal is open or not
  });
});
