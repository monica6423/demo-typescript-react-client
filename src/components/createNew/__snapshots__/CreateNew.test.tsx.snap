// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNew component should render the component 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        class="button-cell "
      >
        <button
          class="button main"
        >
          <b>
            Create new
          </b>
        </button>
        <input
          name="inputField.key"
          placeholder="Search by restaurant, company & type"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
